Salt: support systemd (don't assume Redhat <=> systemd)

justinsb opened the pull request:
Also work around problems with Salt & systemd services, in particular
that Salt doesn't issue a daemon-reload.

This is splitting out the systemd-specific parts of #9381, so that we can focus scrutiny on them (without the AWS distractions).

justinsb committed:
Salt: support systemd (don't assume Redhat <=> systemd)

Also work around problems with Salt & systemd services, in particular
that Salt doesn't issue a daemon-reload.

zmerlynn commented on cluster/saltbase/salt/salt-helpers/services:
```diff
@@ -0,0 +1,67 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set -e
```
Preference is for the normal preamble here:
```
set -o errexit
set -o nounset
set -o pipefail
```

(mostly just for consistency, and readability about `-o errexit` versus `-e`)

zmerlynn commented:
I reviewed this for "doesn't break us" (though Jenkins-bot did a better job there :P). I'm not a great person for final review for systemd competence, though, so don't tag me just because I commented. :)

derekwaynecarr commented:
I will do a review from a systemd perspective.

derekwaynecarr commented on cluster/saltbase/salt/docker/init.sls:
```diff
@@ -114,6 +114,36 @@ lxc-docker-{{ override_docker_ver }}:
+    - defaults:
+        environment_file: {{ environment_file }}
+
+# The docker service.running block below doesn't work reliably
```
It doesn't work for you here because you do not have the docker.service file as a file to watch in the docker.service.running block.  We had a similar problem for the Kubelet in this area that I have a fix for here:
https://github.com/GoogleCloudPlatform/kubernetes/pull/10492

I think you can get rid of fix-service-docker and just add the watch on the file or a require on the file when running in a systemd setup.

justinsb replied:
The root issue that caused me to just move this out of salt is this one: https://github.com/saltstack/salt/issues/16778

Salt won't do a daemon-reload out-of-the-box, by design.  This was not what I was expecting.

I can definitely add the extra watch, but it won't work if e.g. docker has already been loaded by systemd.

derekwaynecarr replied:
Interesting.  I wonder if I didnt encounter this before because Fedora came with a .service file as part of the yum install of docker.

derekwaynecarr commented on cluster/saltbase/salt/docker/docker.service:
```diff
@@ -0,0 +1,20 @@
+StartLimitInterval=0
+
+[Install]
+WantedBy=multi-user.target
```
So in a Fedora 21 setup, I do get a packaged file from the repositories.

```
[vagrant@kubernetes-master ~]$ cat /usr/lib/systemd/system/docker.service
[Unit]
Description=Docker Application Container Engine
Documentation=http://docs.docker.com
After=network.target

[Service]
Type=notify
EnvironmentFile=-/etc/sysconfig/docker
EnvironmentFile=-/etc/sysconfig/docker-storage
EnvironmentFile=-/etc/sysconfig/docker-network
Environment=GOTRACEBACK=crash
ExecStart=/usr/bin/docker -d \
          $OPTIONS \
          $DOCKER_STORAGE_OPTIONS \
          $DOCKER_NETWORK_OPTIONS \
          $INSECURE_REGISTRY
LimitNOFILE=1048576
LimitNPROC=1048576
LimitCORE=infinity
MountFlags=slave

[Install]
WantedBy=multi-user.target
```

Can we align your sample file with the one provided by that distro?

justinsb replied:
I'd be happy to, except where did this come from (i.e. is it GPL?)

derekwaynecarr replied:
http://pkgs.fedoraproject.org/cgit/docker.git/tree/docker.service

derekwaynecarr replied:
@justinsb - So the behavior differences are not major concern right now for the Vagrant setup, I think I am fine if you keep the file you have now.  I am not running with any custom storage solutions now, so I don't think I need the ease of the separate environment files.

derekwaynecarr replied:
@justinsb - or better, in case someone does a yum update, can you only write this file if the os_family is not RedHat since the RedHat packaged docker does include a service file?  so basically, this is the only os specific issue among systemd distros right now.

justinsb replied:
I think that's a good idea; definitely at this stage we want to minimizing changes!

I think this whole thing is in the `else` block of `{% if grains.os_family == 'RedHat' %}`, so I think we get the (more complete) docker service as provided by the OS on RedHat, and we only have to install our own on non-Redhat systemd (i.e. Jessie and Vervet).

derekwaynecarr commented on cluster/saltbase/salt/kube-addons/init.sls:
```diff
@@ -101,13 +101,17 @@ addon-dir-create:
+{% if pillar.get('is_systemd') %}
 
-/usr/lib/systemd/system/kube-addons.service:
+{{ pillar.get('systemd_system_path') }}/kube-addons.service:
```
i think the same holds here, can kube-addons service.running just have a require on this file existing?

justinsb replied:
I will add; I don't think it is sufficient because of the daemon-reload salt issue, but I agree it is better to declare dependencies.

derekwaynecarr commented on cluster/saltbase/salt/salt-helpers/services:
```diff
@@ -0,0 +1,67 @@
+#!/bin/bash
```
So I find this easier to reason than Salt, but I think the issue that you were encountering is that the service.running clauses in the salt files did not have a dependency to run after the .service file is written to disk.  I think if we just add the file watch for the .service files uniformly, you won't need this.

derekwaynecarr commented:
If you can not write an alternative docker.service file, then I think I am lgtm on this PR.

Ping when updated, and will test.

justinsb committed:
Set standard bash options in services helper script

justinsb committed:
Salt: Add more dependencies to systemd services

This may help Salt reload services correctly, although we still
need the script until Salt's bug with reloading services on systemd
is resolved.

Salt bug: https://github.com/saltstack/salt/issues/16778

justinsb commented:
Updated with suggestions, although I didn't try to replace Redhat's Docker systemd service file.  @derekwaynecarr Are you OK with this now?  I want to make it super-low risk, so if there's anything else I can do for that, just LMK.

justinsb committed:
Salt: clean daemon_args on RedHat, not systemd

I don't think it actually makes a difference based on how we're using it,
but this is now more of a precise change.

justinsb committed:
Salt: Add service block for kube-proxy for systemd

We have the workaround for Salt's problems here, but we should still express
our intent.

justinsb committed:
Salt: have kube-addons service depend on init file

For symmetry with systemd

justinsb committed:
Salt: re-remove comment block

It had been removed in upstream, but had been kept by mistake here.

justinsb commented:
(Ouch, so either the e2e tests were a flake before, or Salt can't cope with empty -watch blocks.)

justinsb commented:
@derekwaynecarr any chance you can have a look at this?  It is essentially blocking AWS.  cluster/kube-up does work with vagrant - LMK if there is more I can be testing!

thockin commented:
Anything else blocking this?  I know this is pretty important for AWS support, but it's SO close to the wire...

benmccann commented:
It'd be really rough not to have AWS working with 1.0 :-/  I'm not super familiar with Salt, but this looks mostly like just a refactoring and I imagine this chunk of code should get used enough that it'll suss out any issues relatively quickly

thockin commented:
I'm waiting for a status update...

On Mon, Jul 6, 2015 at 4:49 PM, Ben McCann <notifications@github.com> wrote:

> It'd be really rough not to have AWS working with 1.0 :-/ I'm not super
> familiar with Salt, but this looks mostly like just a refactoring and I
> imagine this chunk of code should get used enough that it'll suss out any
> issues relatively quickly
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/10488#issuecomment-119029709>
> .
>

derekwaynecarr commented:
I don't want to block this, but I can't test until tomorrow as I have been
on vacation.

On Monday, July 6, 2015, Tim Hockin <notifications@github.com> wrote:

> I'm waiting for a status update...
>
> On Mon, Jul 6, 2015 at 4:49 PM, Ben McCann <notifications@github.com
> <javascript:_e(%7B%7D,'cvml','notifications@github.com');>> wrote:
>
> > It'd be really rough not to have AWS working with 1.0 :-/ I'm not super
> > familiar with Salt, but this looks mostly like just a refactoring and I
> > imagine this chunk of code should get used enough that it'll suss out any
> > issues relatively quickly
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <
> https://github.com/GoogleCloudPlatform/kubernetes/pull/10488#issuecomment-119029709
> >
> > .
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/10488#issuecomment-119030004>
> .
>

justinsb commented:
@derekwaynecarr I can run the vagrant tests.  I guess I should run the e2e test suite.  Let me know if there is anything else I should run!

justinsb commented:
Here are the results:

```
Summarizing 5 Failures:

[Fail] Kubectl client Simple pod [It] should support port-forward
/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/kubectl.go:197

[Fail] ServiceAccounts [It] should mount an API token into pods
/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/util.go:869

[Fail] Services [It] should release the load balancer when Type goes from LoadBalancer -> NodePort
/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/service.go:528

[Fail] MaxPods [BeforeEach] Validates MaxPods limit number of pods that are allowed to run.
/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/max_pods.go:59

[Fail] Services [It] should be able to change the type and nodeport settings of a service
/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/service.go:399

Ran 48 of 92 Specs in 4396.237 seconds
FAIL! -- 43 Passed | 5 Failed | 2 Pending | 42 Skipped --- FAIL: TestE2E (4401.44s)
FAIL
```

I don't think vagrant supports load-balancers; MaxPods and mount-API token is known-flaky I believe, and port-forward relies on socat being installed (https://github.com/GoogleCloudPlatform/kubernetes/pull/10712).

justinsb commented:
So I think the e2e tests look good for vagrant; e2e tests for GCE look good; e2e tests for AWS look good.

Can we merge?  I can't guarantee that this doesn't introduce new problems, but I don't think we will find those problems until we merge.  It is very late, I know (sorry!).  As it is so late, an alternative would be to run AWS from a copy of cluster/saltbase with this patches (either in-tree or in my own tree, I would rather it be in-tree and we just build two salt packages).  Any other de-risking suggestions gratefully received!

derekwaynecarr commented:
Vagrant won't work until this merges

https://github.com/GoogleCloudPlatform/kubernetes/pull/10562

I don't want to hold this up for you.  I can rebase my PR and ensure that
things are good tomorrow with whatever gets in repository at that time.

On Tuesday, July 7, 2015, Justin Santa Barbara <notifications@github.com>
wrote:

> @derekwaynecarr <https://github.com/derekwaynecarr> I can run the vagrant
> tests. I guess I should run the e2e test suite. Let me know if there is
> anything else I should run!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/10488#issuecomment-119205757>
> .
>

thockin commented:
So do we run with this?  The risk adds up by the hour.  @zmerlynn thoughts?  How scared are you and is it more or less scary than AWS not working?

zmerlynn commented:
I reviewed this a while ago from a "probably not going to break core" perspective. The problem now is that it just missed the last release train before 1.0, so it's going to miss a lot of external exposure prior to 1.0 shipping, so even if we merge it and GCE shakes out, there's some chance we'll miss some weird broken case. That risk seems acceptable to take for AWS support, and something we could cherry-pick in post 1.0, given that @justinsb has more or less given us continuous support, and "occasional unknown community complaints" are just a vague boogieman.

thockin commented:
Agree.  I am giving this LGTM and kicking shippable.

@bgrant0607 or @brendandburns for #2

justinsb commented:
OK, thanks!  Not having a public version pre-release is definitely scary, but I will definitely be available to figure out any weirdness that happens.

Are we going to have a conformance-test-party or something like that?

thockin commented:
Are you coming to Portland? :)

On Tue, Jul 7, 2015 at 3:54 PM, Justin Santa Barbara <
notifications@github.com> wrote:

> OK, thanks! Not having a public version pre-release is definitely scary,
> but I will definitely be available to figure out any weirdness that happens.
>
> Are we going to have a conformance-test-party or something like that?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/10488#issuecomment-119366763>
> .
>

justinsb commented:
Yes, I'll be there all day Tuesday!

thockin commented:
Then Tuesday is our conformance test hack session :)

On Tue, Jul 7, 2015 at 4:17 PM, Justin Santa Barbara <
notifications@github.com> wrote:

> Yes, I'll be there all day Tuesday!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/10488#issuecomment-119372148>
> .
>

justinsb commented:
Ping?  I feel like it is now or never (before v1) on this one...

thockin commented:
Still 50% love from shippable.  I'm going to OK this.

@vmarmol - this is a scary one - watch closely, please.

@zmerlynn  I guess this is a cherry pick

vmarmol commented:
Merging, lets keep our eyes peeled for this one.

justinsb commented:
Awesome; re-testing on AWS right now

zmerlynn commented:
@thockin: Really? A cherry-pick? One of you can propose it, I gave the instructions. :) (But I suppose we should figure out the review guidelines.)

derekwaynecarr commented:
FYI @justinsb - vagrant appears to come up since this was merged, but it looks like kube2sky is not working for some reason.  I will have to debug later tonight or tomorrow morning on what caused this but appears related to provisioning of the network.

justinsb commented:
@derekwaynecarr Is there an e2e test that is failing?  Or how can I reproduce?

derekwaynecarr commented:
Networking and Services e2e are failing.  My guess is that something is
happening between provision-network.sh and docker install but it looks like
kube2sky is complaining that there is no route to the host.  That was not
the case earlier in the day.  I can't debug for a bit because of kids :-).
Hopefully it's something minor.

On Wednesday, July 8, 2015, Justin Santa Barbara <notifications@github.com>
wrote:

> @derekwaynecarr <https://github.com/derekwaynecarr> Is there an e2e test
> that is failing? Or how can I reproduce?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/10488#issuecomment-119735663>
> .
>

derekwaynecarr commented:
It's possible I had some junk laying around.  I am going to reproduce and
report back.

On Wed, Jul 8, 2015 at 6:42 PM, Derek Carr <decarr@redhat.com> wrote:

> Networking and Services e2e are failing.  My guess is that something is
> happening between provision-network.sh and docker install but it looks like
> kube2sky is complaining that there is no route to the host.  That was not
> the case earlier in the day.  I can't debug for a bit because of kids :-).
> Hopefully it's something minor.
>
>
> On Wednesday, July 8, 2015, Justin Santa Barbara <notifications@github.com>
> wrote:
>
>> @derekwaynecarr <https://github.com/derekwaynecarr> Is there an e2e test
>> that is failing? Or how can I reproduce?
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/GoogleCloudPlatform/kubernetes/pull/10488#issuecomment-119735663>
>> .
>>
>

justinsb commented:
I'm trying too!

derekwaynecarr commented:
@justinsb - I just created a cluster and checked the logs in kube2sky and I
do not appear to have a problem, so its either some type of race, or I had
something laying around previously.  Either way, I will try a few more
times, but it could be a false alarm.  Right now, kube2sky is properly
routing to the host for me.

On Wed, Jul 8, 2015 at 7:29 PM, Justin Santa Barbara <
notifications@github.com> wrote:

> I'm trying too!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/10488#issuecomment-119758546>
> .
>

thockin commented:
When you guys confirm, I'll propose this as a cherry-pick

On Wed, Jul 8, 2015 at 4:39 PM, Derek Carr <notifications@github.com> wrote:

> @justinsb - I just created a cluster and checked the logs in kube2sky and I
> do not appear to have a problem, so its either some type of race, or I had
> something laying around previously. Either way, I will try a few more
> times, but it could be a false alarm. Right now, kube2sky is properly
> routing to the host for me.
>
> On Wed, Jul 8, 2015 at 7:29 PM, Justin Santa Barbara <
> notifications@github.com> wrote:
>
> > I'm trying too!
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <
> https://github.com/GoogleCloudPlatform/kubernetes/pull/10488#issuecomment-119758546
> >
>
> > .
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/10488#issuecomment-119759906>
> .
>

derekwaynecarr commented:
To confirm, I did try this multiple times from HEAD and things appear OK - apologies for earlier confusion.
